numpy/lib/function_base.py

```diff
@@ -898,15 +898,19 @@ def average(a, axis=None, weights=None, returned=False):
     TypeError: Axis must be specified when shapes of a and weights differ.
 
     """
-    if not isinstance(a, np.matrix):
-        a = np.asarray(a)
+    a = np.asanyarray(a)
 
     if weights is None:
         avg = a.mean(axis)
         scl = avg.dtype.type(a.size/avg.size)
     else:
-        a = a + 0.0
-        wgt = np.asarray(weights)
+        wgt = np.asanyarray(weights)
+
+        if issubclass(a.dtype.type, (np.integer, np.bool_)):
```

Comment:
I think the type of `wgt` doesn't matter. The current code does as follows for the four combinations to think about (`float` means `f4` or `f8`):

```
a     wgt        result_dtype
---   ---        ---
int   int        f8
int   float      f8
float int        float
float float2     biggest(float, float2)
```

I suppose in the 2nd line we could cast to whatever wgt's floating type was (f4 or f8) but I'm not sure whether that's helpful or unnecessarily complicated.